server/channels/store/storetest/scheduled_post_store.go

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package storetest
+
+import (
+	"testing"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/shared/request"
+	"github.com/mattermost/mattermost/server/v8/channels/store"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestScheduledPostStore(t *testing.T, rctx request.CTX, ss store.Store, s SqlStore) {
+	t.Run("SaveScheduledPost", func(t *testing.T) { testSaveScheduledPost(t, rctx, ss, s) })
+}
+
+func testSaveScheduledPost(t *testing.T, rctx request.CTX, ss store.Store, s SqlStore) {
+	channel := &model.Channel{
+		TeamId:      "team_id_1",
+		Type:        model.ChannelTypeOpen,
+		Name:        "channel_name",
+		DisplayName: "Channel Name",
+	}
+
+	createdChannel, err := ss.Channel().Save(rctx, channel, 1000)
+	assert.NoError(t, err)
+
+	t.Run("base case", func(t *testing.T) {
+		userId := model.NewId()
+		scheduledPost := &model.ScheduledPost{
+			Draft: model.Draft{
+				CreateAt:  model.GetMillis(),
+				UserId:    userId,
+				ChannelId: createdChannel.Id,
+				Message:   "this is a scheduled post",
+			},
+			ScheduledAt: model.GetMillis() + 100000, // 100 seconds in the future
+		}
+
+		createdScheduledPost, err := ss.ScheduledPost().Save(scheduledPost)
+		assert.NoError(t, err)
+		assert.NotEmpty(t, createdScheduledPost.Id)
+
+		scheduledPostsFromDatabase, err := ss.ScheduledPost().GetScheduledPostsForUser(userId, "team_id_1")
+		assert.NoError(t, err)
+		require.Equal(t, 1, len(scheduledPostsFromDatabase))
+		assert.Equal(t, scheduledPost.Id, scheduledPostsFromDatabase[0].Id)
+	})
+
+	t.Run("scheduling in past should not be allowed", func(t *testing.T) {
```

Comment:
This might just be me, but I find it weird that we test for valid data at the `store` layer rather than the `app` layer. Not sure if we do this across the board, but the store doesn't have any constraints for what the timestamps should be (other than the limitations imposed by the type), so this check feels like it should happen at the `app` layer since it's more of a property of our server logic.

It also means that we could theoretically leverage the store layer to backfill messages for example, if we were migrating for auditing purposes. Not something we'll necessarily do, but it feels more correct to validate in the logic layer vs. the store.

Reply:
I fully agree that business logic validations should always happen on the app layer, but, drafts, posts, and now scheduled posts are an exception.

If you see the IsValid method for these entities, they all require a `maxPostSize` value to check if their message is within the limits of whats allowed on the database. This information - column length of message field, is currently present only in the SQLStore layer. This is why, the validation for these three entities occur on store layer instead of the app layer.

We could split the validation into app and store layer, but I feel its better to consolidate all checks in one place. This also follows the convention we have for posts and drafts.

Reply:
Gotcha, if there's a good reason for this I'm okay with it.